HSEARCH-3088 Query id in DSL

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-3088

I hope it makes sense.
At the moment it only supports strings.

It looks like:
```
		SearchQuery<DocumentReference> query = searchTarget.query()
				.asReference()
				.predicate( f -> f.id().matching( DOCUMENT_1 ).toPredicate() )
				.build();
```

yrodiere commented on engine/src/main/java/org/hibernate/search/engine/search/dsl/predicate/impl/MatchIdPredicateContextImpl.java:
```diff
@@ -0,0 +1,37 @@
+
+	@Override
+	public MatchIdPredicateContext matching(Object value) {
+		matchIdBuilder.value( value );
```
I see you allow multiple values, which is nice; wouldn't be useful to also provide a `matchingAny(Collection<String>)` method? From what I can see it can be implemented easily here with a simple loop, without changing the backend.

DavideD replied:
Sure, sounds good

yrodiere commented on engine/src/main/java/org/hibernate/search/engine/search/predicate/spi/MatchIdPredicateBuilder.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface MatchIdPredicateBuilder<B> extends SearchPredicateBuilder<B> {
+
+	void value(Object values);
```
`values` should be singular here (`value`), I think? Or did I misunderstand how it works?

DavideD replied:
no, it's a typo

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/predicate/impl/ElasticsearchFieldPredicateBuilderFactory.java:
```diff
@@ -29,6 +30,9 @@
 	 */
 	boolean isDslCompatibleWith(ElasticsearchFieldPredicateBuilderFactory other);
 
+	MatchIdPredicateBuilder<ElasticsearchSearchPredicateBuilder> createMatchIdPredicateBuilder(
```
This is unnecessary; you can remove this method and its implementations.

To be clear:

* `SearchPredicateBuilderFactory` is the entry point for the engine to retrieve predicate builders. It's implemented by the backend.
* In some cases, for field-scoped predicates (match, range), `SearchPredicateBuilderFactory` implementations delegate to a field-scoped, backend-specific `ElasticsearchFieldPredicateBuilderFactory` or `LuceneFieldPredicateBuilderFactory`. But that's only when the predicate can have a different behavior depending on the field being targeted: the `match` predicate will accept different input based on whether you target an integer or `LocalDate` field, for example, but this concept doesn't make sense for the `bool` predicate, which doesn't really "target" a field.

So in the case of your ID predicate, you can see that you don't need to change the `ElasticsearchFieldPredicateBuilderFactory`, only the `SearchPredicateBuilderFactory` (which you already did).

I'll try to add some javadoc in a separate PR to make that clearer.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/predicate/impl/LuceneFieldPredicateBuilderFactory.java:
```diff
@@ -32,6 +33,9 @@
 	 */
 	boolean isDslCompatibleWith(LuceneFieldPredicateBuilderFactory other);
 
+	MatchIdPredicateBuilder<LuceneSearchPredicateBuilder> createMatchIdPredicateBuilder(
```
As for the Elasticsearch backend, this is unnecessary; you can remove this method and its implementations.

yrodiere replied:
It seems you didn't take my comment into account? This method and its implementations should be removed.

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/MatchSearchPredicateIT.java:
```diff
@@ -69,6 +69,35 @@ public void setup() {
 	}
 
+	@Test
+	public void match_id() {
```
We tried to create one test class per type of predicate: `MatchSearchPredicateIT`, `RangeSearchPredicateIT`, etc.

Could you please move this to a different test class?

The class can be quite simple, much simpler than this one, because you don't really have to test multiple types of fields. Maybe you could take inspiration from what @fax4ever did here: https://github.com/hibernate/hibernate-search/commit/d0d1c3e1c8e94704451e976f95553cb5071bf694

yrodiere commented:
Follow-up ticket: https://hibernate.atlassian.net/browse/HSEARCH-3416

DavideD commented:
I've applied the remarks and added the method `matchingAny`.

yrodiere commented on engine/src/main/java/org/hibernate/search/engine/search/dsl/predicate/MatchIdPredicateContext.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Collection;
+
+/**
+ * The context used when starting to define a match predicate.
```
Just noticed this: this is apparently a copy/paste from the match predicate, it should be rephrased or removed.

DavideD commented:
Should be fixed now.

I've updated the javadoc and removed the method I forgot to remove before (I did it for elasticsearch but I forgot to do it for Lucene).

yrodiere commented:
Merged, thanks!
